src/Rules/Preg/RegexPatternRule.php

```diff
@@ -0,0 +1,119 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Rules\Preg;
+
+use PhpParser\Node;
+use PhpParser\Node\Expr\FuncCall;
+use PHPStan\Analyser\Scope;
+use PHPStan\Type\Constant\ConstantArrayType;
+use PHPStan\Type\Constant\ConstantStringType;
+
+class RegexPatternRule implements \PHPStan\Rules\Rule
+{
+
+	public function getNodeType(): string
+	{
+		return FuncCall::class;
+	}
+
+	/**
+	 * @param FuncCall $node
+	 * @param \PHPStan\Analyser\Scope $scope
+	 * @return string[]
+	 */
+	public function processNode(Node $node, Scope $scope): array
+	{
+		$patterns = $this->extractPatterns($node, $scope);
+
+		$errors = [];
+		foreach ($patterns as $pattern) {
+			if ($this->validatePattern($pattern)) {
+				continue;
+			}
+
+			$errors[] = sprintf('The regex pattern \'%s\' is not valid.', $pattern);
+		}
+
+		return $errors;
+	}
+
+	/**
+	 * @param FuncCall $functionCall
+	 * @param Scope $scope
+	 * @return string[]
+	 */
+	private function extractPatterns(FuncCall $functionCall, Scope $scope): array
+	{
+		if (!isset($functionCall->args[0])) {
+			return [];
+		}
+		$patternNode = $functionCall->args[0]->value;
+		$patternType = $scope->getType($patternNode);
+
+		if (!$functionCall->name instanceof Node\Name) {
+			return [];
+		}
+		$functionName = (string) $functionCall->name;
+
+		if (
+			$patternType instanceof ConstantStringType
+			&& in_array($functionName, [
+				'preg_match',
+				'preg_match_all',
+				'preg_split',
+				'preg_grep',
+				'preg_replace',
+				'preg_replace_callback',
+				'preg_filter',
+			], true)
+		) {
+			return [$patternType->getValue()];
+		}
+
+		if (
+			$patternType instanceof ConstantArrayType
```

Comment:
$patternType can be a union of constant arrays, use `TypeUtils::getConstantArrays` and adjust accordingly.

Reply:
It can also theoretically be an `ArrayType` with constant string values.

Reply:
> It can also theoretically be an ArrayType with constant string values.

How does this one work? I couldn't create such a type from code.

Reply:
I managed to do it: https://phpstan.org/r/251b6b2d4094c16f0759b01c67957ca3

Reply:
@ondrejmirtes Awesome, thanks!